gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

```diff
@@ -3880,6 +4027,54 @@ private Object getOldVersionedEntry(LocalRegion dataRegion, Object key, RegionEn
     }
   }
 
+  // Writer should add old entry with tombstone with region version in the common map
+  // wait till writer has written to common old entry map.
+  private static Object getOldVersionedEntry(TXState tx,LocalRegion dataRegion, Object key, RegionEntry re) {
+
+    Object oldEntry = dataRegion.getCache().readOldEntry(dataRegion, key, tx.getCurrentSnapshot(),
+        true, re, tx);
+    if (oldEntry != null) {
+      return oldEntry;
+    } else {
+      // wait till it is populated..
+      // The update/destroy guy can update the region version first and then modify the RE
+      // later copy it to running tx so that when tx misses the entry it is sure that
+      // it will be copied by writer thread
+      // If we copy first and then update the region version and RE then there is a window where
+      // concurrent tx can miss the old entry.
+      // 1. Copy of the old value
+      // 2. New tx starts and takes the snapshot
+      // 3. old tx increments the regionVersion
+      // 4. New tx scans and misses the changed RE as its version is higher than the snapshot.
+      // 5. old tx changes the RE
+
+      // For Transaction NONE we can get locally. For tx isolation level RC/RR
+      // we will have to get from a common DS.
+      oldEntry = dataRegion.getCache().readOldEntry(dataRegion, key, tx.getCurrentSnapshot(), true, re, tx);
+      int numtimes = 0;
+      while (oldEntry == null) {
+        if (TXStateProxy.LOG_FINE) {
+          LogWriterI18n logger = dataRegion.getLogWriterI18n();
+          logger.info(LocalizedStrings.DEBUG, " Waiting for older entry for this snapshot to arrive " +
+              "for key " + key + " re " + re + " for region " + dataRegion.getFullPath());
+        }
+        try {
+          //TODO: Suranjan Should we wait indefinitely? or throw warning and return the current entry.
+          // As
+          if (numtimes < 10) {
```

Comment:
Wow. This might cause unnecessary delay. When getOldVersionedEntry method is called, i.e. this method, the caller is sure that it will find it in the oldEntryMap. Can there be a situation that this unnecessarily keeps on waiting. Second thing is that 10 ms looks too long for waiting for a local op to populate the map. May be we can start with couple and increase to 10 later ( instead of 100 later )